onnxruntime/test/ir/flatbuffer_utils_test.cc

```diff
@@ -0,0 +1,395 @@
+  }
+
+  ASSERT_EQ(output_stream.tellp(), 840) << "Data written to the external file is incorrect.";
+  output_stream.close();
+  ASSERT_TRUE(output_stream.good()) << "Failed to close data file.";
+
+  auto fbs_tensors_offset = builder.CreateVector(fbs_tensors);
+  fbs::test::TestDataBuilder tdb(builder);
+  tdb.add_initializers(fbs_tensors_offset);
+  builder.Finish(tdb.Finish());
+  auto fb_data = builder.GetBufferSpan();
+
+  auto test_data = fbs::test::GetTestData(fb_data.data());
+  auto fbs_tensors2 = test_data->initializers();
+
+  // read
+  std::ifstream input_stream;
+  ExternalDataReader reader;
+  CreateReader("ExternalWriteReadBasicTest.bin", input_stream, reader);
+  static onnxruntime::CPUExecutionProviderInfo info;
+  static onnxruntime::CPUExecutionProvider cpu_provider(info);
+  AllocatorPtr cpu_allocator = cpu_provider.CreatePreferredAllocators()[0];
+
+  std::vector<Tensor> loaded_tensors;
+
+  for (const auto* fbs_tensor : *fbs_tensors2) {
+    Tensor ort_tensor;
+    std::string fbs_tensor_name = fbs_tensor->name()->str();
+    ASSERT_STATUS_OK(LoadOrtTensorOrtFormat(*fbs_tensor, cpu_allocator, fbs_tensor_name, ort_tensor, reader));
+    loaded_tensors.push_back(std::move(ort_tensor));
+  }
+
+  bool data_validated = true;
+
+  ASSERT_EQ(initializers.size(), loaded_tensors.size());
+
+  // convert expected initializers (TensorProtos) to Tensors for easier comparison
+  std::vector<Tensor> expected_tensors;
+  const Env& env = Env::Default();
+  const wchar_t* placeholder_model_path = L"placeholder_model_path";
+
+  for (int i = 0; i < static_cast<int>(initializers.size()); i++) {
+    auto expected_proto = initializers[i];
+    TensorShape tensor_shape = utils::GetTensorShapeFromTensorProto(expected_proto);
+    const DataTypeImpl* const type = DataTypeImpl::TensorTypeFromONNXEnum(expected_proto.data_type())->GetElementType();
+    Tensor expected_tensor(type, tensor_shape, cpu_allocator);
+    ASSERT_STATUS_OK(utils::TensorProtoToTensor(env, placeholder_model_path, initializers[i], expected_tensor));
+    expected_tensors.push_back(std::move(expected_tensor));
+  }
+
+  // validate data
+  for (int i = 0; i < static_cast<int>(expected_tensors.size()); i++) {
+    auto& expected_tensor = expected_tensors[i];
+    auto& loaded_tensor = loaded_tensors[i];
+    ASSERT_EQ(expected_tensor.DataType(), loaded_tensor.DataType());
+    ASSERT_EQ(expected_tensor.Shape(), loaded_tensor.Shape());
+    ASSERT_EQ(expected_tensor.SizeInBytes(), loaded_tensor.SizeInBytes());
+    std::vector<uint8_t> expected_data(static_cast<uint8_t*>(expected_tensor.MutableDataRaw()),
+                                       static_cast<uint8_t*>(expected_tensor.MutableDataRaw()) + expected_tensor.SizeInBytes());
+    std::vector<uint8_t> loaded_data(static_cast<uint8_t*>(loaded_tensor.MutableDataRaw()),
+                                     static_cast<uint8_t*>(loaded_tensor.MutableDataRaw()) + loaded_tensor.SizeInBytes());
```

Comment:
Use const wherever possible. As we're not modifying the data, we can use DataRaw() instead.
 
Use gsl::span wherever possible. It can be used with a range of different container types so is a flexible type that we use in a lot of APIs. Also avoids copying data here.